File: src/main/java/com/pascalnb/dbwrapper/ConfigAuthenticator.java
package com.pascalnb.dbwrapper;

import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.util.Properties;

/**
 * Implementation of [DatabaseAuthenticator] that reads the credentials from a cfg file.
 *
 * @see DatabaseAuthenticator
 */
class ConfigAuthenticator extends DatabaseAuthenticator {

    private final String[] credentialsArray;

    protected ConfigAuthenticator(String configFile) {
        Properties properties = new Properties();
        try {
            File jarPath = new File(getClass().getProtectionDomain().getCodeSource().getLocation().getPath());
            String propertiesPath = jarPath.getParentFile().getAbsolutePath();
            properties.load(new FileInputStream(propertiesPath + "/" + configFile));
        } catch (FileNotFoundException __) {
            try (InputStream config = getClass().getClassLoader().getResourceAsStream(configFile)) {
                properties.load(config);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }

        String username = properties.getProperty("username");
        String password = properties.getProperty("password");
        String url = properties.getProperty("host");

        try {
            Class.forName(properties.getProperty("driver"));
        } catch (ClassNotFoundException e) {
            throw new DatabaseException("JDBC driver not loaded");
        }

        this.credentialsArray = new String[]{username, password, url};
    }

    /**
     * Returns the database credentials from the credentials file.
     */
    @NotNull
    @Override
    protected String @NotNull [] getCredentials() {
        return credentialsArray;
    }

}

File: src/main/kotlin/com/pascalnb/dbwrapper/ConfigAuthenticator.kt
package com.pascalnb.dbwrapper

import java.io.*
import java.util.*

/**
 * Implementation of [DatabaseAuthenticator] that reads the credentials from a cfg file.
 *
 * @see DatabaseAuthenticator
 */
@Suppress("unused")
internal class ConfigAuthenticator(configFile: String) : DatabaseAuthenticator() {

    private val credentialsArray: Array<String?>

    init {
        val properties = Properties()
        try {
            val jarPath = File(javaClass.protectionDomain.codeSource.location.path)
            val propertiesPath = jarPath.parentFile.absolutePath
            properties.load(FileInputStream("$propertiesPath/$configFile"))
        } catch (e: FileNotFoundException) {
            try {
                javaClass.classLoader.getResourceAsStream(configFile).use { config -> properties.load(config) }
            } catch (e2: IOException) {
                throw UncheckedIOException(e2)
            }
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }

        val username = properties.getProperty("username")
        val password = properties.getProperty("password")
        val url = properties.getProperty("host")

        try {
            Class.forName(properties.getProperty("driver"))
        } catch (e: ClassNotFoundException) {
            throw DatabaseException("JDBC driver not loaded")
        }

        credentialsArray = arrayOf(username, password, url)
    }

    /**
     * Returns the database credentials from the credentials file.
     */
    override val credentials: Array<String?> get() = credentialsArray
}